rpcs3/rpcs3qt/update_manager.cpp

```diff
@@ -0,0 +1,516 @@
+}
+
+bool update_manager::handle_reply(QNetworkReply* reply, std::function<bool(update_manager& man, const QByteArray&, bool)> func, bool automatic, const std::string& message)
+{
+	if (auto error = reply->error(); error != QNetworkReply::NoError)
+	{
+		reply->deleteLater();
+		if (error == QNetworkReply::OperationCanceledError)
+		{
+			QString error = reply->errorString();
+			LOG_ERROR(GENERAL, "[Auto-updater] Network Error: %s", error.toStdString());
+		}
+		return false;
+	}
+
+	// Read data from network reply
+	const QByteArray data = reply->readAll();
+	reply->deleteLater();
+
+	LOG_NOTICE(GENERAL, "[Auto-updater] %s", message);
+	if (!func(*this, data, automatic))
+	{
+		if (!automatic)
+			QMessageBox::warning(m_progress_dialog, tr("Auto-updater"), tr("An error occured during the auto-updating process.\nCheck the log for more information."));
+
+		m_progress_dialog->close();
+	}
+
+	return true;
+}
+
+bool update_manager::handle_json(const QByteArray& data, bool automatic)
+{
+	const QJsonObject json_data = QJsonDocument::fromJson(data).object();
+	int return_code             = json_data["return_code"].toInt(-3);
+
+	bool hash_found = true;
+
+	if (return_code < 0)
+	{
+		std::string error_message;
+		switch (return_code)
+		{
+		case -1: error_message = "Hash not found(Custom/PR build)"; break;
+		case -2: error_message = "Server Error - Maintenance Mode"; break;
+		case -3: error_message = "Server Error - Return code not found"; break;
+		default: error_message = "Server Error - Unknown Error"; break;
+		}
+
+		if (return_code != -1)
+			LOG_ERROR(GENERAL, "[Auto-updater] error: %s return code: %d", error_message, return_code);
+		else
+			LOG_WARNING(GENERAL, "[Auto-updater] error: %s return code: %d", error_message, return_code);
+
+		// If a user clicks "Check for Updates" with a custom build ask him if he's sure he wants to update to latest version
+		if (!automatic && return_code == -1)
+		{
+			if (QMessageBox::question(m_progress_dialog, tr("Auto-updater"), tr("You're currently using a custom build.\nDo you want to update to the latest official rpcs3 version?"),
+			        QMessageBox::Yes | QMessageBox::No) == QMessageBox::No)
+			{
+				m_progress_dialog->close();
+				return true;
+			}
+
+			hash_found = false;
+		}
+		else
+		{
+			return false;
+		}
+	}
+
+	const auto& latest = json_data["latest_build"];
+	if (!latest.isObject())
+	{
+		LOG_ERROR(GENERAL, "[Auto-updater] Json doesn't contain latest_build section");
+		return false;
+	}
+
+	QString os;
+
+#ifdef _WIN32
+	os = "windows";
+#elif defined(__linux__)
+	os = "linux";
+#else
+	LOG_ERROR(GENERAL, "[Auto-updater] Your os isn't currently supported by the auto-updater");
+	return false;
+#endif
+
+	if (!latest[os].isObject() || !latest[os]["download"].isString() || !latest["version"].isString() || !latest[os]["size"].isDouble() || !latest[os]["checksum"].isString())
+	{
+		LOG_ERROR(GENERAL, "[Auto-updater] Some information seems unavailable", os.toStdString());
```

Comment:
Am I missing something here? I think the %s is missing.

Reply:
Ah originally it was a message saying info about os was unavailable but I just started checking a lot of things instead, changed the message and forgot to remove the parameter.